Apply selective malli mu.fn enforcement for certain namespaces to improve performance

escherize opened the pull request:
resolves #37144 

This PR is a step that I think will solve the perf issue. My approach has been to modify how schema validation functions behave, particularly in our production environment. Here's a summary of what I've done:

- Selective Schema Validation: I've augmented the macro to selectively disable the emission of schema validation code for certain namespaces from `mu.fn/fn` and `mu/defn`. This was done to reduce cpu load, especially validations that might not be critical in production

- Different Behavior in Production: The changes I made ensure that these schema validations are skipped when the system is in a production environment, as indicated by config/is-prod?. The idea is to enhance performance where it matters most to our users.

- Maintaining a List of Namespaces: I've put together a list of namespaces for which these changes should apply, but it is incomplete. This targeted approach allows us to mostly maintain the rigor of validations in prod while avoiding bottlenecks.

- Balancing Risk and Performance: Any changes like this carry some risk, so I've tried to be cautious, limiting the scope of these changes to reduce the possibility of unintended effects on other parts of our system.

Our primary goal has been to address the concerns raised about high CPU usage and slow loading times in the latest Metabase version. By not running any schema validation in production, I'm hopeful that we'll see improvements in these areas.

---

Another thing this does is skip printing a message we were suspicious of causing some heavy cpu usage (in prod only).

escherize committed:
rough approach for per-namespace schema checking

escherize committed:
Added tests

escherize committed:
cleanup + add a namespace to skip

escherize commented on src/metabase/util/malli/fn.clj:
```diff
@@ -295,8 +295,40 @@
+;; FIXME: is there a cleaner way to setup `namespaces-toskip`?
+;; I was thinking we could `swap!` it from the namespaces themselves with:
+;; `(swap namespaces-toskip conj *ns*)`
+(require ['metabase.lib.metadata.calculation])
+
+(def namespaces-toskip
+  "Used to track namespaces to not enforce malli schemas on with `mu.fn/fn`."
+  (atom #{(find-ns 'metabase.lib.metadata.calculation)}))
```
Open to suggestions about a clean way to do this, as well as which namespaces we should toskip.

piranha replied:
Maybe a metadata on namespace directly? Like `^:malli/skip`, for example?

bshepherdson replied:
Yeah, I think decorating the namespaces makes sense. On the other hand we probably want this for all of `metabase.lib.*` and that's a lot of files to remember to do that in. Though I guess we could use a linter to require either `^:malli/skip` or `^:malli/enforce` within `metabase.lib.*`.

dpsutton replied:
I don't think I would operate on namespaces as objects, just on the symbol or string of the name. Have the set of namespaces be symbols or regexes rather than `clojure.lang.Namespace` objects.

calherries commented on test/metabase/util/malli/fn_skip_test.clj:
```diff
@@ -0,0 +1,54 @@
+        (finally
+          (reset! mu.fn/namespaces-toskip original-nts))))))
+
+(comment
+
+  (clojure.test/run-tests)
+
+  )
```
```suggestion
```

bshepherdson commented on src/metabase/util/malli/fn.clj:
```diff
@@ -344,7 +378,10 @@
-                        {})]
-    (instrumented-fn-form error-context (parse-fn-tail fn-tail))))
+  (let [parsed (parse-fn-tail fn-tail)]
+    `(if (#'*skip-ns-decision-fn*)
```
If I'm reading this right, this is compiling the `if` into the code where the function is being constructed. That doesn't seem ideal... but there doesn't seem to be a way to check at macro expansion time.

I suppose it's not so bad if it's ultimately a top-level thing like a `mu/defn`.

escherize replied:
Yeah, I didn't start with this approach, but there doesn't seem to be a way to check at macro-expansion time. 

Pointing out the obvious, but the `if` is evaluated at `mu.fn/fn` compile-time, so any extra time to run it will occur during fn creation, and not fn execution.

dpsutton replied:
I think we have everything we need at macro-expansion time right? The only input is the namespace which is known at compile time. I think that condition can be outside of emitted code.

escherize replied:
I couldn't figure out how to get it working with re-binding or alter-var-rooting or swap!ing the state required to test it without emitting the if. Maybe I missed something there?

escherize replied:
I got the emission (and altering behavior in tests) working properly, with help from @dpsutton and @noahmoss 😎.

escherize committed:
use stringified namespaces instead of ns objects

dpsutton commented on src/metabase/util/malli/fn.clj:
```diff
@@ -295,8 +295,38 @@
+;; FIXME: is there a cleaner way to setup `namespaces-toskip`?
+;; I was thinking we could `swap!` it from the namespaces themselves with:
+;; `(swap namespaces-toskip conj *ns*)`
+(def namespaces-toskip
```
something like this might be better:

```clojure
(defn instrument-ns?
  "Used to track namespaces to not enforce malli schemas on with `mu.fn/fn`."
  [namespace]
  (let [lib-and-middleware [#"^metabase\.lib\..*"
                            #"^metabase\.query-processor\.middleware\..*"]
        matches?    (fn [namespace regexes]
                      (let [n (-> namespace ns-name str)]
                        (some #(re-matches % n) regexes)))
        ad-hoc      '#{metabase.pulse.render} ;; just to demonstrate
        m           (meta namespace)]
    (or (:instrument/always m)
        (true? (:instrument/never m))
        (not (matches? namespace lib-and-middleware))
        (not (contains? ad-hoc (ns-name namespace))))))
```

escherize committed:
move to `instrument-ns?`+ trying to not emit an if

dpsutton commented on src/metabase/util/malli/fn.clj:
```diff
@@ -312,16 +312,29 @@
+
+(def ^:dynamic *skip-ns-decision-fn*
+  (core/fn skip-ns-decision-fn [namespace]
+    (and #_config/is-prod? config/is-dev? ;;nocommit
```
we need to remove the nocommit and switch from `is-dev?` to `is-prod?`

escherize replied:
done

dpsutton commented on src/metabase/util/malli/fn.clj:
```diff
@@ -312,16 +312,29 @@
+    (and #_config/is-prod? config/is-dev? ;;nocommit
+         (let [should? (instrument-ns? namespace)]
+           (when (not should?)
+             (println "skipping instrumenting var in " (ns-name namespace)))
```
println needs to go

escherize replied:
done

escherize committed:
WIP, working on tests

escherize committed:
eval

dpsutton committed:
Fixup tests and remove some errant junk

tests are a bit tricky because the the reader expands the macro so we
can't do anything at runtime to affect the binding. This is why we use
eval so we can coordinate when the binding happens

dpsutton committed:
some notes and remove the pulse ns from ad-hoc exclusions

escherize committed:
add docstrings and comments

escherize commented on src/metabase/util/malli/defn.clj:
```diff
@@ -71,10 +71,17 @@
+            :cljs (mu.fn/deparameterized-fn-form parsed)))
+      `(def ~(vary-meta fn-name merge attr-map)
+         ~docstring
+         ~(macros/case
```
this can just be `(mu.fn/deparameterized-fn-form parsed)`.

escherize replied:
done

escherize committed:
remove redundant case statement

escherize committed:
swapped branches in mu/defn, so that when `skip?` we emit deparameterized fn

escherize committed:
skip logging message about calculating display name in prod

- it was suspected as part of the high cpu usage issue here:
  https://github.com/metabase/metabase/issues/37144

escherize committed:
config/is-prod? is not avaliable in cljs

escherize committed:
use the right kind of quote

escherize committed:
actually it's ok to use binding if we delay the evaluation of the macro

escherize committed:
use :cljs-dev and :cljs-release to skip printing sus message in prod

escherize committed:
use cljs-dev and cljs-release to control printing message

escherize committed:
add more tests for defn, and move fn_skip_test into fn_test

noahmoss reviewed (approved):
Macro changes make sense to me now that I'm more up to speed on the implementation. My one question is should the docstring for `mu/defn` be updated to indicate that there's a way to exclude certain namespaces/ns patterns from validation? Otherwise it might be a little hard to discover unless you dig into the macro code

escherize committed:
add info in mu/defn docstring about instrument-ns? skipping behavior

escherize committed:
fix test
